Support windows.

sighingnow opened the pull request:
+ Use `_write` in printString.
+ Use `<io.h>` on windows instead of `<unistd.h>`.
+ `llvm-config`, `llc`, `opt` and `llvm-link` are not available on windows, use `clang` to compile the LLVM code and link as executable.

sighingnow commented:
I have encounter another problem on windows. When compile `.ll` file, I got the following error:

~~~shell
D:\Local\Temp\sixten21340\Sixten.Builtin.ll:2198:14: error: '%is-inited' defined with type 'i1'
  switch i1* %is-inited, label %not-inited [i1 1, label %inited]
             ^
1 error generated.
~~~

I have check the llvm code in `Sixten.Builtin.ll`:

~~~
2195 > define external fastcc void @Sixten.Builtin-init(){
2196 > ; <label>:0:
2197 >   %is-inited = load i1, i1* @Sixten.Builtin-inited, align 1
2198 >   switch i1* %is-inited, label %not-inited [i1 1, label %inited]
2199 > inited:
2200 >   ret void
2201 > not-inited:
~~~

After modify `i1* %is-inited` as `i1 %is-inited`, I can make it compiled successfully and then get the normal executable. I'm not familiar with the llvm codegen part, @ollef  could your please have a look at this problem ?

I'm working with `llvm-hs-pretty-0.1.0.0` and `llvm-hs-pure-5.1.2`.

sighingnow commented:
Clang version:

~~~
>clang --version
clang version 5.0.1 (tags/RELEASE_501/final)
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: D:\Program\LLVM\bin
~~~

ollef commented:
Hey, and welcome!

This looks like a good approach. Do you have any idea why it fails on Travis? It seems to work on my machine.

Here are some more thoughts and comments:

* If we're able to do everything with `clang` on Windows, perhaps we can do so on all platforms? We could then stop relying on `llvm-config` and just optionally take the path to `clang` on the command-line. Do you think that would work?
* Why did the multiline strings need to be changed?
* The codegen problem seems to be that we're relying on a bug fixed in https://github.com/llvm-hs/llvm-hs/pull/171. If you use the version of llvm-hs that's in the currently used Stack snapshot it should work. We'll have to fix this issue once the snapshot is bumped.
* Are there any special installation steps that are needed on Windows?

sighingnow commented:
> If we're able to do everything with clang on Windows, perhaps we can do so on all platforms? We could then stop relying on llvm-config and just optionally take the path to clang on the command-line. Do you think that would work?

It does work, but not perfectly.  I got different binary executables generated via `llvm-link/opt/clang` and only via `clang` on Ubuntu, although they all give the correct output.

> Do you have any idea why it fails on Travis?

After check the stderr log on travis-ci, I found the failure is caused by the `-flto` flags. See [builds/330650266](https://travis-ci.org/sighingnow/sixten/builds/330650266) and [builds/330652767](https://travis-ci.org/sighingnow/sixten/builds/330652767). Now I have enabled the `--catch-stderr` for `stack test` to make the debugging easier.

> Why did the multiline strings need to be changed?

The `CPP` extension will affect multi-line string, having different behavior on windows and linux (the `\\` at the end of line).

> Are there any special installation steps that are needed on Windows?

Only the library [ivmai/bdwgc](https://github.com/ivmai/bdwgc) is required to build manually. I will add more detailed description to README later. Should we add an extra command line flags to allow users to specify where the library file `gc-lib.lib` exists ?

ollef commented on src/Backend/Compile.hs:
```diff
@@ -40,6 +42,22 @@ supportedLlvmVersions = makeVersion . (: [minorVersion]) <$> supportedMajorVersi
+-- We simply assume that @clang.exe@ already exists in @%PATH%@.
+--
+clangBinPath :: String -> IO FilePath
+clangBinPath command = trim <$> checkClangExists trySuffixes
```
Perhaps inline the argument since it's only called with `"clang"`?

sighingnow replied:
Fixed.

ollef commented on src/Backend/Compile.hs:
```diff
@@ -40,6 +42,22 @@ supportedLlvmVersions = makeVersion . (: [minorVersion]) <$> supportedMajorVersi
+  where trySuffixes = "" : fmap (('-' :) . showVersion) supportedLlvmVersions
+        checkClangExists (suffix : xs) =
+          handle (\(_ :: IOException) -> checkClangExists xs)
+          $ readProcess (command ++ suffix) ["-print-prog-name=" ++ command ++ suffix] ""
```
Do Clang installations for Windows add a version suffix like the Ubuntu packages? (The binary is called e.g. `clang-5.0`.)

Otherwise this might be redundant work.

sighingnow replied:
Fixed.

ollef commented:
I've updated the master branch to Stackage LTS 10.3 and fixed the codegen errors you were seeing. So Sixten should now work with llvm-hs 5.1.3, however older versions will not be supported anymore.

Let me know if I can help with anything here. I'd love to get this merged. :)

sighingnow commented:
I have rebased this PR to master.

> What was it that didn't work perfectly?

The binary generated via `llvm-link/opt/clang` and only via `clang` are different, although they all give the correct output. The reason for the difference needs further investigation and I still keep the `llvm-link/opt/clang` way for non-windows system. I think this issue shouldn't block the work of windows support.

> It might make sense to add an -L flag for specifying library paths in general, which would help with locating gc-lib.

I have added the `--extra-lib-dir` option for this proposal.

ollef commented on src/Command/Compile/Options.hs:
```diff
@@ -9,4 +9,5 @@ data Options = Options
   , optimisation :: Maybe String
   , assemblyDir :: Maybe FilePath
   , llvmConfig :: Maybe FilePath
+  , extraLibDir :: Maybe FilePath
```
Can this be a `[FilePath]` instead so we can specify multiple paths? I think you can use `many` instead of `optional` in the command-line parser to make that work.

sighingnow replied:
Fixed. Thanks for your suggestions!

ollef commented:
🎉
